### PR TITLE
Feature/auto lock

### PR DESCRIPTION
**Pull request by OlegGordiichuk**
- Code clean
- Add new setting
- Add new logic for auto lock 
- Fix tests

### CODE REVIEW

File: Trust/Settings/ViewControllers/SettingsViewController.swift [hunk 1]
```diff
@@ -32,9 +39,37 @@ class SettingsViewController: FormViewController, Coordinator {
...
+                return
+            }
+            $0.title = strongSelf.viewModel.autoLockTitle
+            $0.options = strongSelf.viewModel.autoLockOptions
+            $0.value = strongSelf.lock.getAutoLockType()
+            $0.selectorTitle = strongSelf.viewModel.autoLockTitle
+            $0.displayValueFor = { value in
+                return value?.displayName
+            }
+            $0.hidden = Condition.function(["PasscodeRow"], { form in
```
Comment (vikmeup) [inline Trust/Settings/ViewControllers/SettingsViewController.swift]: can you move `PasscodeRow ` into constants? there Values struct

File: Trust/Settings/ViewControllers/SettingsViewController.swift [hunk 2]
```diff
@@ -32,9 +39,37 @@ class SettingsViewController: FormViewController, Coordinator {
...
+                return !((form.rowBy(tag: "PasscodeRow") as? SwitchRow)?.value ?? false)
```
Comment (vikmeup) [inline Trust/Settings/ViewControllers/SettingsViewController.swift]: can you add this option properly? take a look at examples above

File: Trust/Settings/ViewControllers/SettingsViewController.swift [hunk 3]
```diff
@@ -11,17 +11,24 @@ protocol SettingsViewControllerDelegate: class {
...
+
     struct Values {
         static let currencyPopularKey = "0"
         static let currencyAllKey = "1"
     }
+
     private var config = Config()
+
     private var lock = Lock()
+
```
Comment (vikmeup) [inline Trust/Settings/ViewControllers/SettingsViewController.swift]: no need to add spaces

File: Trust/Lock/Lock.swift [hunk 4]
```diff
@@ -6,73 +6,147 @@ import KeychainSwift
...
         dateFormatter.timeStyle = DateFormatter.Style.short
         return dateFormatter
     }
+
+    private func autoLockTriggered() -> Bool {
+        let type = getAutoLockType()
+        switch type {
+        case .disabled:
+            return true
+        default:
```
Comment (vikmeup) [inline Trust/Lock/Lock.swift]: can you make it exhaustive?

File: Trust/Lock/Lock.swift [hunk 5]
```diff
@@ -6,73 +6,147 @@ import KeychainSwift
...
     private struct Keys {
         static let service = "trust.lock"
         static let account = "trust.account"
     }
+
     private let passcodeAttempts = "passcodeAttempts"
+
     private let maxAttemptTime = "maxAttemptTime"
+
+    private let autoLockType = "autoLockType"
```
Comment (vikmeup) [inline Trust/Lock/Lock.swift]: no need to add extra spaces

File: Trust/Settings/ViewModels/SettingsViewModel.swift [hunk 6]
```diff
@@ -25,6 +25,16 @@ struct SettingsViewModel {
         ]
     }
 
+    var autoLockOptions: [AutoLock] {
+        return [
+            AutoLock.disabled,
+            AutoLock.oneMinute,
+            AutoLock.fiveMinutes,
+            AutoLock.oneHour,
+            AutoLock.fiveHours
```
[the remark this answers is not in the document]
Comment (OlegGordiichuk) [reply Trust/Settings/ViewModels/SettingsViewModel.swift]: From telegram communication I will add "Immediate".

[the remark this answers is not in the document]
Comment (OlegGordiichuk) [reply Trust/Settings/ViewModels/SettingsViewModel.swift]: Done

File: TrustTests/Factories/FakeLockProtocol.swift [hunk 7]
```diff
@@ -4,8 +4,16 @@ import UIKit
 @testable import Trust
 
 class FakeLockProtocol: LockInterface {
+
     var passcodeSet = true
+
+    var showProtection = true
+
```
[the remark this answers is not in the document]
Comment (OlegGordiichuk) [reply TrustTests/Factories/FakeLockProtocol.swift]: Fixed

File: Trust/Settings/ViewControllers/SettingsViewController.swift (hunk 3 above)
Comment (OlegGordiichuk) [reply Trust/Settings/ViewControllers/SettingsViewController.swift]: Done

File: Trust/Lock/Lock.swift (hunk 5 above)
Comment (OlegGordiichuk) [reply Trust/Lock/Lock.swift]: Done

File: Trust/Lock/Coordinators/LockEnterPasscodeCoordinator.swift [hunk 8]
```diff
@@ -22,14 +22,14 @@ class LockEnterPasscodeCoordinator: Coordinator {
...
     func start() {
-        guard lock.isPasscodeSet() else { return }
+        guard lock.shouldShowProtection() else { return }
 
         window.rootViewController = lockEnterPasscodeViewController
         window.makeKeyAndVisible()
     }
-
+    //This method should be refactored!!!
     func showAuthentication() {
```
[the remark this answers is not in the document]
Comment (OlegGordiichuk) [reply Trust/Lock/Coordinators/LockEnterPasscodeCoordinator.swift]: In this step we will show auth steps only when we already has 
presented lock screen.

File: TrustTests/Coordinators/LockEnterPasscodeCoordinatorTest.swift [hunk 9]
```diff
@@ -32,4 +32,15 @@ class LockEnterPasscodeCoordinatorTest: XCTestCase {
         coordinator.start()
         XCTAssertTrue(coordinator.window.isHidden)
     }
+
+    func testDisabledAutoLock() {
```
[the remark this answers is not in the document]
Comment (OlegGordiichuk) [reply TrustTests/Coordinators/LockEnterPasscodeCoordinatorTest.swift]: Add as separate issue https://github.com/TrustWallet/trust-wallet-ios/issues/716

**Pull request merged**